sgkit/stats/hwe.py

```diff
@@ -0,0 +1,178 @@
+    ----------
+    - [1] Wigginton, Janis E., David J. Cutler, and Goncalo R. Abecasis. 2005.
+        “A Note on Exact Tests of Hardy-Weinberg Equilibrium.” American Journal of
+        Human Genetics 76 (5): 887–93.
+
+    Raises
+    ------
+    ValueError
+        If any observed counts are negative.
+    """
+    if obs_hom1 < 0 or obs_hom2 < 0 or obs_hets < 0:
+        raise ValueError("Observed genotype counts must be positive")
+
+    obs_homc = obs_hom2 if obs_hom1 < obs_hom2 else obs_hom1
+    obs_homr = obs_hom1 if obs_hom1 < obs_hom2 else obs_hom2
+    obs_mac = 2 * obs_homr + obs_hets
+    obs_n = obs_hets + obs_homc + obs_homr
+    het_probs = np.zeros(obs_mac + 1, dtype=np.float64)
+
+    if obs_n == 0:
+        return np.nan  # type: ignore[no-any-return]
+
+    # Identify distribution midpoint
+    mid = int(obs_mac * (2 * obs_n - obs_mac) / (2 * obs_n))
+    if (obs_mac & 1) ^ (mid & 1):
+        mid += 1
+    het_probs[mid] = 1.0
+    prob_sum = het_probs[mid]
+
+    # Integrate downward from distribution midpoint
+    curr_hets = mid
+    curr_homr = int((obs_mac - mid) / 2)
+    curr_homc = obs_n - curr_hets - curr_homr
+    while curr_hets > 1:
+        het_probs[curr_hets - 2] = (
+            het_probs[curr_hets]
+            * curr_hets
+            * (curr_hets - 1.0)
+            / (4.0 * (curr_homr + 1.0) * (curr_homc + 1.0))
+        )
+        prob_sum += het_probs[curr_hets - 2]
+        curr_homr += 1
+        curr_homc += 1
+        curr_hets -= 2
+
+    # Integrate upward from distribution midpoint
+    curr_hets = mid
+    curr_homr = int((obs_mac - mid) / 2)
+    curr_homc = obs_n - curr_hets - curr_homr
+    while curr_hets <= obs_mac - 2:
+        het_probs[curr_hets + 2] = (
+            het_probs[curr_hets]
+            * 4.0
+            * curr_homr
+            * curr_homc
+            / ((curr_hets + 2.0) * (curr_hets + 1.0))
+        )
+        prob_sum += het_probs[curr_hets + 2]
+        curr_homr -= 1
+        curr_homc -= 1
+        curr_hets += 2
+
+    if prob_sum <= 0:  # pragma: no cover
+        return np.nan  # type: ignore[no-any-return]
+    het_probs = het_probs / prob_sum
+    p = het_probs[het_probs <= het_probs[obs_hets]].sum()
+    p = max(min(1.0, p), 0.0)
+
+    return p  # type: ignore[no-any-return]
+
+
+# Benchmarks show ~25% improvement w/ fastmath on large (~10M) counts
+hardy_weinberg_p_value_jit = njit(hardy_weinberg_p_value, fastmath=True)
+
+
+def hardy_weinberg_p_value_vec(
+    obs_hets: ndarray, obs_hom1: ndarray, obs_hom2: ndarray
+) -> ndarray:
+    arrs = [obs_hets, obs_hom1, obs_hom2]
+    if len(set(map(len, arrs))) != 1:
+        raise ValueError("All arrays must have same length")
+    if list(set(map(lambda x: x.ndim, arrs))) != [1]:
+        raise ValueError("All arrays must be 1D")
+    n = len(obs_hets)
+    p = np.empty(n, dtype=np.float64)
+    for i in range(n):
+        p[i] = hardy_weinberg_p_value_jit(obs_hets[i], obs_hom1[i], obs_hom2[i])
```

Comment:
Observation: we're using Dask for parallelization here (across blocks) which is fine, but we may want to consider numba's `parallel=True` in the future for this loop as another dispatch option.